source/reduce/remove_selection_reduction_opportunity_finder.cpp

```diff
@@ -0,0 +1,68 @@
+// Copyright (c) 2019 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "remove_selection_reduction_opportunity_finder.h"
+#include "remove_selection_reduction_opportunity.h"
+
+namespace spvtools {
+namespace reduce {
+
+using namespace opt;
+
+std::string RemoveSelectionReductionOpportunityFinder::GetName() const {
+  return "RemoveSelectionReductionOpportunityFinder";
+}
+
+std::vector<std::unique_ptr<ReductionOpportunity>>
+RemoveSelectionReductionOpportunityFinder::GetAvailableOpportunities(
+    opt::IRContext* context) const {
+  std::vector<std::unique_ptr<ReductionOpportunity>> result;
+
+  // Find all opportunities for removing a selection construct by branching to
+  // the LHS of the construct before finding corresponding opportunities for
+  // branching to the RHS of the construct, to reduce the extent to which
+  // reduction opportunities that can disable one another are adjacent.
+  for (auto choose_lhs : {true, false}) {
+    // Consider every function.
+    for (auto& function : *context->module()) {
+      // Consider every block in the function.
+      for (auto& block : function) {
+        // Check whether the block heads a selection construct.
+        auto merge_inst = block.GetMergeInst();
+        if (!merge_inst || merge_inst->opcode() != SpvOpSelectionMerge) {
```

Comment:
Given code like:
```
%1 = OpLabel
OpSelectionMerge %merge None
OpBranchConditional %cond %2 %3
%2 = OpLabel
OpBranch %merge
%3 = OpLabel
OpBranchConditional %cond2 %merge %4
%4 = OpLabel
OpBranch %merge
```
I feel like the reducer will produce an invalid cfg. Because of the "break" in %3 there is no selection merge instruction, but if you remove the selection merge in %1, you would now need one in %3.

Reply:
Interesting counterexample, thanks.

My thought on how to go forward is to have the pass look for OpBranchConditional instructions for which one target is the conditional's merge block, and to change each to be just an OpBranch to the target that is not the merge block.

I believe that this should work because the only case where an OpBranchConditional inside a selection construct can be intimately tied to the selection construct has this form.  If an OpBranchConditional doesn't target the selection construct's merge block, then removing the selection construct cannot make that OpBranchConditional invalid (it might jump to the merge block or continue target of a loop, say, but that's fine as the loop isn't being removed).

What do you think?

Reply:
So for my example you would change the branch in %3 to just go to %4 if you remove the merge instruction in %1? That sounds look it should work.

Reply:
Yes.  I'll adapt the PR accordingly.

Reply:
OK, done.